### PR TITLE
Servlet 6.1 Factory

### LINKED ISSUE #26806: Servlet 6.1 Factory
Initial Servlet 6.1 factory project

### PR DESCRIPTION
**Pull request by pmd1nh**
Create the factory for Servlet 6.1

fixes #26806

### CODE REVIEW

**Comment by pmd1nh** [comment]
#build
        #spawn.fullfat.buckets=io.openliberty.webcontainer.servlet.6.0.internal_fat,com.ibm.ws.webcontainer.servlet.6.0_fat_toleration.lWAS,com.ibm.ws.webcontainer_fat,com.ibm.ws.webcontainer.servlet.3.1_fat,com.ibm.ws.webcontainer.servlet.4.0_fat,com.ibm.ws.webcontainer_fat_part2,com.ibm.ws.webcontainer_fat_servlet31toleration,com.ibm.ws.webcontainer_fat_servlet31toleration_part2,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0_part2,com.ibm.ws.webcontainer_fat_lWAS,com.ibm.ws.webcontainer_fat_servlet31_webcontainerlWAS,com.ibm.ws.webcontainer.servlet.4.0_fat_toleration.3.0.lWAS

**Comment by pmd1nh** [comment]
fixed #26806

**Comment by pmd1nh** [comment]
Name | Tests | Failures | Errors | Time Taken | Run Order
-- | -- | -- | -- | -- | --

no component | 98630 | 7 | 1 |   |  
com.ibm.ws.wsat_fat.ssl.misroute FAT tests | 6 | 6 | 0 |   |  
io.openliberty.microprofile.opentracing.3.0.internal_fat_tck FAT tests | 69 | 0 | 1 |   |  
io.openliberty.security.jakartasec.3.0.internal_fat.config.2 FAT tests | 76 | 1 | 0 |   |  
io.openliberty.security.jakartasec.fat.config.tests.ConfigurationSigningTests | 7 | 1 | 0 | 2 m 44 s | 7
ConfigurationSigningTests_providerSignsWithRS256_clientExpectsRS256_shortJwksConnectionTimeout_withELOverride_jwt | 1 | 1 | 0 | 2 m 14 s | 5

File: dev/com.ibm.ws.webcontainer.servlet.4.0/src/com/ibm/ws/webcontainer40/osgi/webapp/WebApp40.java [hunk 1]
```diff
@@ -76,7 +73,7 @@ public WebApp40(WebAppConfiguration webAppConfig,
     @Override
     public int getMajorVersion() {
         int level = WebContainer.getServletContainerSpecLevel();
-        return level == WebContainer.SPEC_LEVEL_60 ? 6 : level == WebContainer.SPEC_LEVEL_50 ? 5 : 4;
+        return level == WebContainer.SPEC_LEVEL_61 ? 61 : level == WebContainer.SPEC_LEVEL_60 ? 6 : level == WebContainer.SPEC_LEVEL_50 ? 5 : 4;
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer.servlet.4.0/src/com/ibm/ws/webcontainer40/osgi/webapp/WebApp40.java]: I think this is a type and SPEC_LEVEL_60 || SPEC_LEVEL_61 the major version should be 6.

Comment (pnicolucci) [reply dev/com.ibm.ws.webcontainer.servlet.4.0/src/com/ibm/ws/webcontainer40/osgi/webapp/WebApp40.java]: We also will need an update for the minor version.

File: dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/container/config/WebAppConfiguratorHelper.java [hunk 2]
```diff
@@ -240,7 +240,9 @@ public static boolean allowDefaultErrorPageInServlet30() {
      */
     public static int getVersionId(String version) throws IllegalStateException {
         int versionID = 0;
-        if ("6.0".equals(version)) {
+        if ("6.1".equals(version)) {
```
Comment (pnicolucci) [inline dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/container/config/WebAppConfiguratorHelper.java:243]: Is this reading the version from the web.xml?

File: dev/io.openliberty.webcontainer.servlet.6.1.internal.factories/bnd.bnd [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  io.openliberty.webcontainer.servlet.internal.async.factory.AsyncContextFactory61Impl, \
+  io.openliberty.webcontainer.servlet.internal.osgi.container.config.factory.WebAppConfiguratorFactory61Impl, \
+  io.openliberty.webcontainer.servlet.internal.osgi.response.factory.IRequestResponseFactory61Impl, \
+  io.openliberty.webcontainer.servlet.internal.osgi.srt.factory.SRTConnectionContextPool61Impl, \
+  io.openliberty.webcontainer.servlet.internal.osgi.webapp.factory.WebAppFactory61Impl, \
+  io.openliberty.webcontainer.servlet.internal.servlet.factory.CacheServletWrapperFactory61Impl, \
+  io.openliberty.webcontainer.servlet.internal.session.impl.factory.SessionContextRegistryImplFactory61Impl, \
+  io.openliberty.webcontainer.servlet.internal.util.factory.URIMatcherFactory61Impl
+
+#Need to update to v61 ???
```
Comment (pnicolucci) [inline dev/io.openliberty.webcontainer.servlet.6.1.internal.factories/bnd.bnd]: Where is this used today? I know we use the properties file within the WC components. Can you look to see where this is used to determine if we need an update?

**Comment by pnicolucci** [comment]
@pmd1nh are you going to update the Servlet 6.1 private feature to point to this factory bundle so it is used when the servlet-6.1 public feature is used?

**Comment by pmd1nh** [comment]
> @pmd1nh are you going to update the Servlet 6.1 private feature to point to this factory bundle so it is used when the servlet-6.1 public feature is used?

updated internal-6.1.feature

File: dev/com.ibm.ws.webcontainer.servlet.4.0/src/com/ibm/ws/webcontainer40/osgi/webapp/WebApp40.java (hunk 1 above)
Comment (pmd1nh) [reply dev/com.ibm.ws.webcontainer.servlet.4.0/src/com/ibm/ws/webcontainer40/osgi/webapp/WebApp40.java:76]: WebContainer.getServletContainerSpecLevel() will return an int from 

_dev/io.openliberty.webcontainer.servlet.6.1.internal.factories/src/com/ibm/ws/webcontainer/speclevel/servletSpecLevel.properties_

File: dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/container/config/WebAppConfiguratorHelper.java (hunk 2 above)
Comment (pmd1nh) [reply dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/osgi/container/config/WebAppConfiguratorHelper.java:243]: yes

File: dev/com.ibm.ws.webcontainer.servlet.4.0/src/com/ibm/ws/webcontainer40/osgi/webapp/WebApp40.java (hunk 1 above)
Comment (pmd1nh) [reply dev/com.ibm.ws.webcontainer.servlet.4.0/src/com/ibm/ws/webcontainer40/osgi/webapp/WebApp40.java]: Updated WebApp40.  If we have a new WebApp60 later on, I will move these 2 getter version to the new class.

File: dev/io.openliberty.webcontainer.servlet.6.1.internal.factories/bnd.bnd (hunk 3 above)
Comment (pmd1nh) [reply dev/io.openliberty.webcontainer.servlet.6.1.internal.factories/bnd.bnd]: updated

File: dev/io.openliberty.webcontainer.servlet.6.1.internal.factories/bnd.bnd [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+javac.source: 17
+javac.target: 17
+
+Require-Capability: osgi.ee; filter:="(&(osgi.ee=JavaSE)(version=17))"
+
+Bundle-Name: WAS WebContainer
+Bundle-SymbolicName: io.openliberty.webcontainer.servlet.6.1.internal.factories
+Bundle-Description: WAS WebContainer with object factories for Servlet 6.1 support
+WLP-ServerName: IBM WebSphere Application Server
+WLP-ServerVersion: ${liberty.beta.version}
```
Comment (pnicolucci) [inline dev/io.openliberty.webcontainer.servlet.6.1.internal.factories/bnd.bnd]: Can you open an issue for these two WLP-Server* directives? I only see these in some WebTier bundles so I'd like to clean them up if they are not necessary any longer. Just associate the issue with the Servlet Epic.

**Review by pnicolucci — COMMENTED** [review]
If you're happy with the .settings in: https://github.com/OpenLiberty/open-liberty/pull/26842 you can review that for me and add those here to this PR :)

File: dev/io.openliberty.webcontainer.servlet.6.1.internal.factories/bnd.bnd (hunk 4 above)
Comment (pmd1nh) [reply dev/io.openliberty.webcontainer.servlet.6.1.internal.factories/bnd.bnd]: https://github.com/OpenLiberty/open-liberty/issues/26859

**Comment by pmd1nh** [comment]
#build

File: dev/io.openliberty.webcontainer.servlet.6.1.internal.factories/bnd.bnd (hunk 4 above)
Comment (pmd1nh) [reply dev/io.openliberty.webcontainer.servlet.6.1.internal.factories/bnd.bnd]: Checked with Jared.  These 2 can be removed.

**Comment by pmd1nh** [comment]
#build

**Comment by pmd1nh** [comment]
@pnicolucci all are updated.  Can you please re-review? Thanks!

**Comment by pmd1nh** [comment]
#libby

**Review by pnicolucci — APPROVED** [review]

**Comment by pmd1nh** [comment]
#build

**Comment by pmd1nh** [comment]
Only 2 FAT failure but unrelated
com.ibm.ws.springboot.fat20_fat FAT tests
	
1
	
0
	
1
	
	
io.openliberty.springboot.fat30_fat FAT tests
	
1
	
0
	
1

**Pull request merged**